mcv/patchwork.py

```diff
@@ -0,0 +1,157 @@
+"""Patchwork"""
+import json
+import os
+import os.path
+import subprocess
+from functools import partial
+
+import requests
+
+PATCHWORK_API_ENDPOINT = os.getenv('PATCHWORK_API_ENDPOINT',
+                                   "https://api.patchworksecurity.com/api/v1")
+DEFAULT_UUID_PATH = '/etc/patchwork/uuid'
+DPKG_QUERY_CMD = ["/usr/bin/dpkg-query", "-W",
+                  "-f", "${Status}\t${Package}\t${Version}\n"]
+
+
+def register(api_key, name, uuid_path=DEFAULT_UUID_PATH):
+    """Register this machine
+
+    UUID metadata is stored in uuid_path. The machine is only
+    registered if the UUID metadata isn't found.
+
+    Returns:
+        dict: JSON metadata for this machine
+    """
+    if os.path.exists(uuid_path):
+        with open(uuid_path, 'r') as f:
+            machine_data = json.load(f)
+
+        return machine_data
+
+    _mkdirs(uuid_path)
+
+    os_data = _get_os_info()
+    os_data.update({'name': name})
+
+    url = "/".join([
+        PATCHWORK_API_ENDPOINT,
+        "machine"])
+
+    machine_data = _post(api_key=api_key, url=url, data=os_data)
+    _save_uuid(uuid_path, machine_data)
+
+    return machine_data
+
+
+def update(api_key, packages, uuid=None, uuid_path=DEFAULT_UUID_PATH):
+    """Update this machine's package set
+
+    This replaces the package set stored on the server
+
+    Returns:
+        dict: The JSON output from the server
+    """
+    if uuid is None:
+        uuid = _get_uuid(uuid_path)
+
+    data = []
+
+    for status, package, version in packages:
+        # only upload packages in the 'ii' state otherwise there
+        # may be false positives
+        if status == 'install ok installed':
+            data.append({
+                'name': package,
+                'version': version,
+            })
+
+    url = "/".join([
+        PATCHWORK_API_ENDPOINT,
+        "machine",
+        uuid])
+
+    return _post(api_key=api_key, url=url, data=data)
+
+
+def gather_packages():
```

Comment:
Ideally I think this would go in `mcv.apt` (and then one day you'd need to figure out how to generalize it across apt, yum, etc.)